### ISSUE TITLE
Add `emmett-core` parser

### ISSUE THREAD

**Issue by gi0baro** [defnull/multipart_bench#1]
Hey 👋 
Would be nice to see https://github.com/emmett-framework/core added to the benchmark suite :)

I can open up a PR for that if that's ok.
The only question I have is wether treat is as _Blocking_ or _Not blocking_, as `emmett-core` parser will always store parts in memory and file parts on temporary files.

**Comment by defnull** [comment]
I already wrote the tests for emmett, but I'm not sure what to do with the results yet. I did not include it yet for two reasons:

* It's a pure rust implementation, all the others are pure python or mostly python. Adding compiled languages to the mix would turn this parser benchmark into a language benchmark. The language overhead would completely dominate the results, and hide what this benchmark was actually intended to show: How different implementation approaches behave in different scenarios. And, to be honest, how my own parser compares to others.
* Emett has very mixed results and screws with the scale a bit. It is significantly faster (x6) in some tests, so much so that the results of the other parsers are harder to compare. But in the important tests (upload and mixed) it is significantly slower, which is surprising and may even be a bug. I'm not hesitating because it's too fast. It is not. I'm struggling to explain the results and I am not sure if I'm testing it correctly.

I can publish the test cases if you like, so you can have a look. But I would not want to include the results in the README.md for now. Maybe it would make sense to mention Emmett (and its interesting results) in a "honorable mentions" paragraph?

**Comment by gi0baro** [comment]
> It's a pure rust implementation, all the others are pure python or mostly python. Adding compiled languages to the mix would turn this parser benchmark into a language benchmark. The language overhead would completely dominate the results, and hide what this benchmark was actually intended to show: How different implementation approaches behave in different scenarios. And, to be honest, how my own parser compares to others.

I understand your point, but at the same time I would ask: would you make the same observation if `emmett-core` would've been written in Cython or C? To be completely fair: the implementation is written in rust but uses PyO3 library to inter-operate with the Python interpreter (and not directly with `pyo3-ffi`), so there's still quite a lot of runtime overhead (type checks, casting and conversions, error handling, etc.) made by PyO3 compared to a pure Rust implementation.
To me the existence and availability of compiled python packages is well known and accepted by the Python community, I can think of several super-popular packages in that regards (NumPY, Polars, uvloop, ujson, orjson, etc.), and I don't think the general perception is some *language war*.
But I totally understand if you don't want to mix pure Python implementation vs everything else.

> Emett has very mixed results and screws with the scale a bit. It is significantly faster (x6) in some tests, so much so that the results of the other parsers are harder to compare. But in the important tests (upload and mixed) it is significantly slower, which is surprising and may even be a bug. I'm not hesitating because it's too fast. It is not. I'm struggling to explain the results and I am not sure if I'm testing it correctly.

If you can push the code to a branch/draft PR I would be happy to take a look and debug the whole thing locally.
Given the slow down seems to be completely related to the presence of file parts, to me it can be either:
- the test is different on `emmett-core` vs the other ones: as I said `emmett-core` always write to disk, whereas I'm not sure about all the other packages. If that's the case, we can either: exclude it from file parts tests or I can update the package to support a custom non-file writer to better support the test
- there's something wrong in `emmett-core` in regards to file descriptors or buffers; in that case I would be super-happy to debug it locally and eventually fix the issue

> I can publish the test cases if you like, so you can have a look. But I would not want to include the results in the README.md for now. Maybe it would make sense to mention Emmett (and its interesting results) in a "honorable mentions" paragraph?

That makes sense. An alternative idea might be to add non-pure Python implementation as separated tables, where you compare those just with the fastest pure Python implementation (your package, probably?). Let me know wdyt :)

**Comment by defnull** [comment]
Sorry for the long wait, I wanted to give some libraries a chance to fix some bugs before publishing a new worstcase scenario. That took a while, and then I forgot about it ;) 

Instead of separate tables, I decided to make `multipart` the baseline (100%). It's the fastest pure-python parser in all tests, so this ensures that the pure-python parsers are still easy to compare. All have a score between 0% and 100%.

I also added `emmett-core` as a normal competitor. If a Cython parser is in there, then a Rust parser should also be allowed. I'll try to explain strange or noteworthy results below each test, which gives some context to the very mixed results for `emmett-code`. If you find a bug in the test case, or fix a performance bug in emmett, then I'm happy to re-run all tests and publish new results.
